File: csharp/Tachyon.Sdk/Collection.cs
namespace Tachyon.Sdk;

/// <summary>
/// A handle scoped to one collection. Get one via <see cref="TachyonClient.Collection"/>;
/// it does not verify the collection exists until you make a request.
/// </summary>
public sealed class Collection
{
    private readonly HttpTransport _transport;

    public string Name { get; }
    public DocumentsResource Documents { get; }

    internal Collection(HttpTransport transport, string name)
    {
        _transport = transport;
        Name = name;
        Documents = new DocumentsResource(transport, name);
    }

    /// <summary><c>GET /collections/{name}/search</c>.</summary>
    public Task<SearchResponse> SearchAsync(SearchParams? searchParams = null, CancellationToken cancellationToken = default)
    {
        searchParams ??= new SearchParams();
        var query = new Dictionary<string, string?>
        {
            ["q"] = searchParams.Q,
            ["query_by"] = Join(searchParams.QueryBy),
            ["filter"] = searchParams.Filter,
            ["sort"] = searchParams.Sort,
            ["facet"] = Join(searchParams.Facet),
            ["limit"] = searchParams.Limit?.ToString(),
            ["offset"] = searchParams.Offset?.ToString(),
            ["prefix"] = FormatBool(searchParams.Prefix),
            ["typo_tolerance"] = FormatBool(searchParams.TypoTolerance),
            ["match_mode"] = searchParams.MatchMode switch
            {
                MatchMode.All => "all",
                MatchMode.Any => "any",
                null => null,
                _ => throw new ArgumentOutOfRangeException(nameof(searchParams)),
            },
        };
        return _transport.RequestAsync<SearchResponse>(
            HttpMethod.Get,
            $"/collections/{Uri.EscapeDataString(Name)}/search",
            query: query,
            cancellationToken: cancellationToken);
    }

    /// <summary><c>GET /collections/{name}/suggest</c>.</summary>
    public Task<SuggestResponse> SuggestAsync(SuggestParams suggestParams, CancellationToken cancellationToken = default)
    {
        var query = new Dictionary<string, string?>
        {
            ["q"] = suggestParams.Q,
            ["query_by"] = Join(suggestParams.QueryBy),
            ["limit"] = suggestParams.Limit?.ToString(),
            ["typo_tolerance"] = FormatBool(suggestParams.TypoTolerance),
        };
        return _transport.RequestAsync<SuggestResponse>(
            HttpMethod.Get,
            $"/collections/{Uri.EscapeDataString(Name)}/suggest",
            query: query,
            cancellationToken: cancellationToken);
    }

    private static string? Join(IReadOnlyList<string>? values) =>
        values is null || values.Count == 0 ? null : string.Join(',', values);

    private static string? FormatBool(bool? value) =>
        value is null ? null : value.Value ? "true" : "false";
}

File: java/src/main/java/io/github/adikeshri/tachyon/Collection.java
package io.github.adikeshri.tachyon;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * A handle scoped to one collection. Get one via {@link TachyonClient#collection};
 * it does not verify the collection exists until you make a request.
 */
public final class Collection {
    private final HttpTransport transport;
    private final String name;
    private final DocumentsResource documents;

    Collection(HttpTransport transport, String name) {
        this.transport = transport;
        this.name = name;
        this.documents = new DocumentsResource(transport, name);
    }

    public String name() {
        return name;
    }

    public DocumentsResource documents() {
        return documents;
    }

    /** {@code GET /collections/{name}/search}. */
    public SearchResponse search() {
        return search(SearchParams.empty());
    }

    /** {@code GET /collections/{name}/search}. */
    public SearchResponse search(SearchParams params) {
        Map<String, String> query = new LinkedHashMap<>();
        query.put("q", params.q());
        query.put("query_by", join(params.queryBy()));
        query.put("filter", params.filter());
        query.put("sort", params.sort());
        query.put("facet", join(params.facet()));
        query.put("limit", params.limit() == null ? null : params.limit().toString());
        query.put("offset", params.offset() == null ? null : params.offset().toString());
        query.put("prefix", formatBool(params.prefix()));
        query.put("typo_tolerance", formatBool(params.typoTolerance()));
        query.put("match_mode", params.matchMode() == null ? null : params.matchMode().toString());

        String path = "/collections/" + UrlEncoding.pathSegment(name) + "/search";
        return transport.request("GET", path, query, null, SearchResponse.class);
    }

    /** {@code GET /collections/{name}/suggest}. */
    public SuggestResponse suggest(SuggestParams params) {
        Map<String, String> query = new LinkedHashMap<>();
        query.put("q", params.q());
        query.put("query_by", join(params.queryBy()));
        query.put("limit", params.limit() == null ? null : params.limit().toString());
        query.put("typo_tolerance", formatBool(params.typoTolerance()));

        String path = "/collections/" + UrlEncoding.pathSegment(name) + "/suggest";
        return transport.request("GET", path, query, null, SuggestResponse.class);
    }

    private static String join(List<String> values) {
        return values == null || values.isEmpty() ? null : String.join(",", values);
    }

    private static String formatBool(Boolean value) {
        return value == null ? null : value.toString();
    }
}
